MaxText/tests/check_gpt_vs_reference.py

```diff
@@ -0,0 +1,427 @@
+        "experts.down_proj": torch.randn(config.num_local_experts, config.intermediate_size, config.hidden_size),
+        "experts.down_proj_bias": torch.rand(config.num_local_experts, config.hidden_size),
+    }
+
+    batch_size = 4
+    seq_len = 6
+    hidden_states = torch.randn(batch_size, seq_len, config.hidden_size)
+
+    # reference model
+    model = GptOssMLP(config)
+    model.load_state_dict(predefined_weights)
+    model.eval()
+    with torch.no_grad():
+      expected_output, _ = model(hidden_states)
+
+    # MaxText model
+    cfg = pyconfig.initialize(
+        [None, os.path.join(PKG_DIR, "configs", "base.yml")],
+        run_name="gpt_oss_mlp_test",
+        enable_checkpointing=False,
+        model_name="default",
+        dtype="float32",
+        weight_dtype="float32",
+        megablox=False,
+        sparse_matmul=True,
+        per_device_batch_size=1,
+        max_target_length=seq_len,
+        max_prefill_predict_length=seq_len,
+        base_emb_dim=config.hidden_size,
+        base_mlp_dim=config.intermediate_size,
+        base_moe_mlp_dim=config.intermediate_size,
+        mlp_activations=["sigmoid", "linear"],
+        mlp_activations_limit=config.limit,
+        routed_bias=True,
+        mlp_bias=True,
+        num_experts=config.num_local_experts,
+        num_experts_per_tok=config.num_experts_per_tok,
+        decoder_block="gpt_oss",
+        attention="dot_product",
+    )
+    jax_hidden_states = to_jax(hidden_states)
+    devices_array = maxtext_utils.create_device_mesh(cfg)
+    mesh = Mesh(devices_array, cfg.mesh_axes)
+    jax_model = moe.get_routed_moe(
+        name="MoeBlock",
+        config=cfg,
+        num_experts=cfg.num_experts,
+        num_experts_per_tok=cfg.num_experts_per_tok,
+        mesh=mesh,
+        kernel_init=nd_dense_init(1.0, "fan_in", "truncated_normal"),
+        kernel_axes=("embed", "mlp"),
+        intermediate_dim=cfg.base_mlp_dim,
+        dtype=cfg.dtype,
+    )
+
+    moe_variables = {
+        "moe_variables": {
+            "gate": {
+                "kernel": to_jax(predefined_weights["router.weight"].transpose(0, 1)),
+                "bias": to_jax(predefined_weights["router.bias"]),
+            },
+            "wi_0": to_jax(predefined_weights["experts.gate_up_proj"][..., ::2]),
+            "wi_0_bias": to_jax(predefined_weights["experts.gate_up_proj_bias"][..., ::2]),
+            "wi_1": to_jax(predefined_weights["experts.gate_up_proj"][..., 1::2]),
+            "wi_1_bias": to_jax(predefined_weights["experts.gate_up_proj_bias"][..., 1::2]),
+            "wo": to_jax(predefined_weights["experts.down_proj"]),
+            "wo_bias": to_jax(predefined_weights["experts.down_proj_bias"]),
+        },
+    }
+    actual_output, _ = jax.jit(jax_model.apply)(moe_variables, jax_hidden_states)
+    normalized_expected = to_jax(expected_output) / jnp.linalg.norm(to_jax(expected_output))
+    normalized_actual = actual_output / jnp.linalg.norm(actual_output)
```

Comment:
any reason of comparing normalized values and not unnormalized?

Reply:
~1 seams high, but sometimes with Random weights it's higher than with trained weights.

is it with fp32 or bf16?

Reply:
No, I guess I meant to say, the logits from reference is ~20-30, and delta is about 0.1-2ish, and tolerance is about 1e-1. But our regular logits output usually ~1, not ~20-30. I am using fp32 I think. So after normalizatoin, the tolerance can be 1e-3.